### ISSUE TITLE
extracting character mods as tposes

### ISSUE THREAD

**Issue by russellaugust** [shff/GK3-extractor#2]
I've noticed that the character MOD files come out as OBJs in random stances.  I can only assume your script is either selecting the ANM in alphabetical order or maybe there is a default one.  But they all do seem to have the option for t-poses.  Is there a way to choose which stance they're in?

**Comment by shff** [comment]
@russellaugust Hello!

My script doesn't really touch the ANMs, it just dumps the default position of the MOD files!

I didn't get too far with the ANMs to be fair. Their format is still not 100% clear to me! When I started I thought it used skeletal animation, but Scott Bilas mentioned in the Gamasutra Postmorten that they're storing per-vertex position in the animation files...  I think the other people working in GK3 stuff had more success than I... I should check their implementation when I have more time! 🤔

**Comment by russellaugust** [comment]
thanks @shff I'll chat with some other people who have worked on asset extraction.  I'd guess there's a way to apply a pose or first ANM moment before extraction, but who knows.  G-Engine is quite the mystery.

**Comment by shff** [comment]
@russellaugust Let me know if you get any info on the ANM format! We should probably get together and form a group, there's lots of people interested in GK3 but you're the first person I ever chatted with! 😄

**Comment by russellaugust** [comment]
Agreed! I've had a side project going for the past year with GK3.  I know someone else is working on rebuilding the G-Engine from scratch.  We should all get a group together, since its probably a very small community haha

**Comment by shff** [comment]
@russellaugust I got the ACT parsing to work in the weeken, but still haven’t found a way to connect them to the MODs, but I’ll try to do it in the following days!

After I do this I don’t know if we’ll get T-Poses but it should be possible to at least get different poses!

**Comment by shff** [comment]
Also I got a “merge” mode going that allows to export multiple BSPs, like the entire Rennes Le Chateau map all at once. There’s still some duplicate geometry but it’s cool to walk around the entire city in Blender :)

Btw, would love to know more about your side project! Lemme know if there’s anything I can help!

**Comment by shff** [comment]
@russellaugust Okay, I think I got it 😄 

Now it's able to parse ACT files and extract all its frames into OBJ+MTL pairs, plus the BMP textures (except for t-shirts, for some weird reason!)

Here's an example using `./gk3 GAB_GABPUTTAPE.ACT`:

![Screenshot 2020-10-19 at 23 32 23](https://user-images.githubusercontent.com/25377830/96514069-6d50b480-1263-11eb-8d60-20b803265a62.png)
![Screenshot 2020-10-19 at 23 32 26](https://user-images.githubusercontent.com/25377830/96514073-6de94b00-1263-11eb-80a2-8d5c067ee195.png)

Here's `DEM_DEMTE6STUN.ACT` and `GRA_GRALERREADNOTEA.ACT`

![Screenshot 2020-10-20 at 00 01 50](https://user-images.githubusercontent.com/25377830/96516441-79d70c00-1267-11eb-9756-c1ffaf752d79.png)
![Screenshot 2020-10-20 at 00 02 18](https://user-images.githubusercontent.com/25377830/96516477-88bdbe80-1267-11eb-98b9-906261b03f9d.png)

I'm still getting some mysterious segmentation errors on my machine when accessing arrays, so I had to increase the vertex array size by 20. You might come across those as well... This makes the memory consumption a bit high, but nothing too crazy (the machine I'm using has 2GB of RAM so you're probably safe 😄). Gotta dive into valgrind to fix it sometime. I curse the day I wrote this in C, but gotta live with it!

**Comment by russellaugust** [comment]
this is so awesome, great work!  I have so many questions about what you've figured out.  One big thing is since the OBJs are sort of like an image sequence, does that mean they're being outputted at a frame rate?

**Comment by shff** [comment]
It seems that they're using a fixed rate!

I'm reading every bit in ACT files and there seems to be nothing that could be "frame duration" or anything like that. ANMs text-files also don't have any extra information of the sort in them.

Also, in some animations, some frames have empty "deltas" between them, so I guess the team decided that having them not move was "good enough" and didn't implement any kind of keyframe duration.

My script is a bit wasteful so it just repeats the whole file though, but in the game itself it was probably ok for it to be like this 🤔 

Next steps I'm thinking about doing are trying to export to a proper format with animation like FBX...

Oh, another cool thing: I also found out that the animations aren't 100% vertex-based: they're a mix of vertex based **and** matrix rotation... like, arms and legs ("bones" if you will) are rotated using matrices, and then the vertex-based animation is used to "connect" the bones between themselves. So it *might* be possible to extract the rotations and put them in a rigged model (we're gonna have to rig models ourselves though). 🤔

**Comment by russellaugust** [comment]
This is all really interesting.  

I was wondering about this too, about the ACTs.  If you have the vertex information, it seems possible to invent a rigged bone system.  But it sounds like since the matrix rotation info is there too, it could make it even easier.

**Comment by russellaugust** [comment]
Oh btw, I forgot to answer your earlier question.

My big goal was to see if I could get GK3 running in VR through Unity.  The game already has an unusual interface that actually might work _better_ in VR. Parts like the map geometry puzzle may be too laborious to do in a VR headset but I think being able to explore all the GK3 locations and experience it would be really cool.

Then the larger goal is to start replacing game assets with photogrammetry (since most of the game actually involves real locations).

**Comment by shff** [comment]
@russellaugust Oh wow, seems we have similar ideas, although you're more ambitious! 😉 

I got some reference photography from Rennes le Chateau from last time I was in the area, including some street panoramas (mostly stuff you can't find in Google Images). Lemme know if you're interested. I have friends nearby, in Carcassone so I drop by sometimes too!

A few months ago I was actually trying to redo the whole city using Blender and BlenderGIS, problem is I just don't have the 3d talent or time for it. 😕

**Comment by russellaugust** [comment]
oh wow i am very interested in those photos!  I'm probably better at modeling than I am at programming. I'm pretty good with Python and learned on C++ and JAVA way back in high school and college.  I was going to use this project to learn Unity and C#. Most of my programming skill set is specifically in video and film post production (my day job).
